tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -4939,23 +4939,28 @@ Value *CGMSHLSLRuntime::EmitHLSLLiteralCast(CodeGenFunction &CGF, Value *Src,
     }
   } else if (ConstantFP *CF = dyn_cast<ConstantFP>(Src)) {
     APFloat v = CF->getValueAPF();
-    double dv = v.convertToDouble();
     if (llvm::IntegerType *IT = dyn_cast<llvm::IntegerType>(DstTy)) {
+      APSInt iv;
+      bool isExact;
+      v.convertToInteger(iv, APFloat::roundingMode::rmTowardZero, &isExact);
+      if (IT->getBitWidth() < iv.getBitWidth()) {
+        iv = iv.trunc(IT->getBitWidth());
+      }
       switch (IT->getBitWidth()) {
       case 32:
-        return Builder.getInt32(dv);
+        return Builder.getInt32(iv.getLimitedValue());
```

Comment:
See lines 4914... above for similar code used when casting to integral type.

Reply:
We should specify the actual limits for each of the `getLimitedValue` calls. The default limit is effectively `std::numeric_limits<uint64_t>::max()`. With truncation, this could result in non-saturated values for these casts.

We also should probably add a release note about this since this will absolutely be a behavior change in the compiler.

Reply:
Yes.  For instance, we want to saturate cast to `uint32_t` to between `0` and `std::numeric_limits<uint32_t>::max()` instead of saturating it to signed `int32_t` range.

This will also make it match fxc's behavior for casting floats/doubles.